util/grpc/src/server_cert_reloader.rs

```diff
@@ -0,0 +1,375 @@
+// Copyright (c) 2018-2021 The MobileCoin Foundation
+
+//! A `grpcio::ServerCredentialsFetcher` implementation that reloads a GRPC's server
+//! TLS certificate/key when a SIGHUP is received.
+
+use displaydoc::Display;
+use grpcio::{CertificateRequestType, ServerCredentialsBuilder, ServerCredentialsFetcher};
+use mc_common::logger::{log, Logger};
+use signal_hook::{consts::SIGHUP, flag};
+use std::{
+    fs, io,
+    path::{Path, PathBuf},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+};
+
+/// Certificate Reloader error.
+#[derive(Debug, Display)]
+pub enum ServerCertReloaderError {
+    /// IO: {0}
+    IO(io::Error),
+}
+
+impl From<io::Error> for ServerCertReloaderError {
+    fn from(src: io::Error) -> Self {
+        Self::IO(src)
+    }
+}
+
+/// A `grpcio::ServerCredentialsFetcher` implementation that reloads a GRPC's server
+/// TLS certificate/key when a SIGHUP is received.
+pub struct ServerCertReloader {
+    /// Certificate file to watch.
+    cert_file: PathBuf,
+
+    /// Private ke yfile to watch.
+    key_file: PathBuf,
+
+    /// Signal that we need to re-load the certificate/key files.
+    load_needed: Arc<AtomicBool>,
+
+    /// Logger.
+    logger: Logger,
+}
+
+impl ServerCertReloader {
+    /// Create a new ServerCertReloader that watches `cert_file`/`key_file`.
+    pub fn new(
+        cert_file: &impl AsRef<Path>,
+        key_file: &impl AsRef<Path>,
+        logger: Logger,
+    ) -> Result<Self, ServerCertReloaderError> {
+        let load_needed = Arc::new(AtomicBool::new(true));
+
+        flag::register(SIGHUP, load_needed.clone())?;
+
+        Ok(Self {
+            cert_file: cert_file.as_ref().to_path_buf(),
+            key_file: key_file.as_ref().to_path_buf(),
+            load_needed,
+            logger,
+        })
+    }
+}
+
+impl ServerCredentialsFetcher for ServerCertReloader {
+    fn fetch(&self) -> Result<Option<ServerCredentialsBuilder>, Box<dyn std::error::Error>> {
+        if !self.load_needed.load(Ordering::SeqCst) {
+            return Ok(None);
+        }
+
+        log::info!(self.logger, "Loading certificates");
+
+        let crt = fs::read_to_string(&self.cert_file)?;
+        let key = fs::read_to_string(&self.key_file)?;
+
+        let new_cred = ServerCredentialsBuilder::new()
+            // This sets the client root certificate to verify client's identity.
+            // We are not using this feature, however grpcio still requires something to be set
+            // there when using the ServerCredentialsFetcher mechanism. As a workaround we are
+            // using the server's certificate chain here.
+            .root_cert(
```

Comment:
@jcape I wonder what your thoughts are about this.

Reply:
Are you testing with a self-signed cert that does not have `basicConstraints: CA:FALSE`?

This may fail in a real environment with a proper chain, where the end-user cert is not a CA---i.e. can't possibly sign any client certificates---or grpc changes it's behavior to reject client-auth CAs that aren't CAs.

I'd strongly recommend adding "self-signed" and "ok-rsa-chain" generators to `mc-crypto-x509-test-vectors` and just using those.

Reply:
Does that mean `mc-crypto-x509-test-vectors` would become a dependency of this crate?
I'm fine with that but want to be sure I understand what you're suggesting.

Reply:
`dev-dependency`, yes.

Reply:
Then I don't understand - what would be used for this root_cert in a real environment?

Reply:
Oh, sorry, yes, that would make this a dependency (the other stuff would be dev-dependency).

Reply:
What other stuff?

Reply:
OK, actually, I'm doubly sorry, this is just wrong. test-vectors just reads the files out of build, it's not suitable for a runtime dependency (right now).